asset.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func copyRegularFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	destFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(destFile, srcFile); err != nil {
+		return err
+	}
+	return nil
+}
+
+// makeAbsoluteTarget takes a absolute path which will serve as a base
+// and a relative symlink target to create absolute equivalent of
+// relative target.
+func makeAbsoluteTarget(base, target string) string {
+	dirPart := filepath.Dir(base)
+	testPath := filepath.Join(dirPart, target)
+	return filepath.Clean(testPath)
+}
+
+// absoluteSrcTargetToDest returns an absolute symlink target pointing
+// to an asset on ACI image, which is a counterpart of an absolute
+// symlink target pointing to an asset on local filesystem.
+//
+// Example:
+//
+// Given an asset /assets:$HOME/some/assets and a symlink somewhere
+// inside local asset which points to another place in local asset
+// like:
+// $HOME/some/assets/some/symlink -> $HOME/some/assets/some/target
+// This function returns target /assets/some/target. In this case
+// srcBase is $HOME/some/assets, srcTarget is
+// $HOME/some/assets/some/target and destBase is /assets.
+func absoluteSrcTargetToDest(srcBase, srcTarget, destBase string) (string, error) {
+	relTarget, err := filepath.Rel(srcBase, srcTarget)
+	if err != nil {
+		return "", err
+	}
+	destTarget := filepath.Join(destBase, relTarget)
+	return filepath.Clean(destTarget), nil
+}
+
+// copySymlink copies the symlink, but before doing so it ensures that
+// the symlink is pointing to node inside the asset.
+//
```

Comment:
Hm, could be. Maybe instead of mangling symlink targets and whatnot I should just add a code looking for dangling symlinks after all assets were added and warn about them.